### PR TITLE
WL21484 _- Hotfix - Solve Joint array issues with Similar, but not same Skeletons

### PR DESCRIPTION
**Pull request by Menithal**
Fix for #11290 To Solve Boneitis on Entity Avatars.

## Test Plan in Comments!

This PR Fixes a specific issue with the #11290 Update, where A skeleton similar, but not same in structure got distorted due to bones. This occurred due to the arrays of the Target skeleton and the Animation being different.

Solution was to use BoneIndeces to match the joints between the target and animation.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3047/HighFidelity-Beta-PR11301-119a2703f79d09aeb1e8c7ccacc1ea500cdbc5b5.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3049/HighFidelity-Beta-PR11301-856452405e1145847a4dfdd3f23b286288602fe9.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3049/HighFidelity-Beta-PR11301-856452405e1145847a4dfdd3f23b286288602fe9.exe

**Comment by jherico** [comment]
boneitis

**Comment by Menithal** [comment]
Not  Ready for QA just yet, test plan has not yet been provided! Just verifying if one bug was caused by this or the previous PR before submitting it.

**Comment by Menithal** [comment]
# Test Plan

## Test 1 Cube Test
1. Run the following in the Script console
``` 
var m = "http://norteclabs.com/HF/experiment/PopupTest.fbx";

Entities.addEntity(
{ type: "Model",
  name: "Tester",
  position: Vec3.sum(MyAvatar.position, Vec3.multiplyQbyV(MyAvatar.orientation, {x:0,y:0,z:-2})), 
  modelURL: m,
  animation:{url:m, running: true}});

Entities.addEntity(
{ type: "Model",
  name: "Control",
  position: Vec3.sum(MyAvatar.position, Vec3.multiplyQbyV(MyAvatar.orientation, {x:0,y:0,z:-4})), 
  modelURL: m,
  animation:{url:m, running: true}});
``` 

#### Basic

1. This should create two model entities that consist of two (inverted) boxes, with the top one with a spinning back and forth, and **bobs up and down**.  This tests that by default, the Allow Translation is active if not defined.
2.  Selecting the closest (*Tester*) with CREATE should show a new field under Animation, which is "Animation Allow Translation" which should be enabled. (by default to not break existing content)
3. Disable "Animation Allow Translation" for "Tester". 
4. *Tester*'s top box should **Stop Bobbing up and down, but keep spinning**.  Control should not be affected.
5. Re-enable "Animation Allow Translation"  for "Tester"
6. *Tester*'s top box should **Start bobbing up and down and keep spinning**. Control should not change.

### Export / Import

7. Disable "Animation Allow Translation" for *Tester*
8. Export both *Tester* and *Control* together into to JSON.
9. Delete *Tester* and *Control* from the World
10. Import both *Tester* and *Control* from the JSON
11. *Tester* should still have allowTranslation false, 
12. *Control* should have allowTranslation true.

Above tests import and export steps export the entity date correctly

### Finalization

13. Quit High Fidelity
14. Open High Fidelity again and connect to the same domain the test object are at
15. *Tester* should still have allowTranslation false, 
16. *Control* should have allowTranslation true.

Above tests make sure domain entity settings are consistent.


## Test 2 Rig Test

1. Goto a clear spot in the Domain
2. Run following Script
```
Clipboard.importEntities("https://s3.eu-west-2.amazonaws.com/norteclabs.com/HF/experiment/PR11301.json") ? Clipboard.pasteEntities(Vec3.sum({x:0,z:0,y:-2},MyAvatar.position)) : false

```
3. This will rez two avatars that have three copies.
3. Characters under **Control** should be of different proportion and should be used as reference
4. Characters under **Enabled** have  Enabled Translation. They will look more squished than Control
5. Characters have **Disabled** have Disabled Translation, They should have more coherient proportional animation, and are the main target of this PR.

A character set  will have their bounding box at the same level as others of the same color. however their feet may not in all cases touch the ground, and in some cases may not be aligned due to different in limb length.

Above tests make sure that Entity Rigs now have support regardless of the skeleton. The Red Model has 2 extra bones over the blue one and offsets the bone array. Both Avatars have the same skeleton structure, but have different proportions (other has longer limbs than normal)

### Test 3 Smoke Screen Test

Run Client Test Scripts


## Known Issues 
Following are not under this PR, but noticed in this PR which should be addressed on later PRs and separate tickets.

- Entity Animations ignore animationLoop, lastFrame and currentFrame, So some loops will not work as intended. Seems to also occur in #11250 or even earlier versions. Not in Scope of this PR
- Entity Animations on different proportions may cause the animation to not touch the ground. This is a more advanced topic (as how do you know entity is standing, sitting or running or walking or flying) and not in scope of this PR.

**Comment by Nex-Pro** [comment]
Testing this

**Comment by Nex-Pro** [comment]
In progress awaiting discussion with @Menithal

**Comment by cozza13** [comment]
@Nex-Pro do you have any updates?

**Comment by Nex-Pro** [comment]
I've had a number of crashes ( no splats generated ) when terminating scripts from the qa test menu.

Unhandled exception at 0x00007FFF328234BE (ucrtbase.dll) in interface.exe: Fatal program exit requested. occurred I will check if this is happening in latest dev.

**Comment by Nex-Pro** [comment]
@cozza13 ^

**Comment by Nex-Pro** [comment]
Call Stack.

>	ucrtbase.dll!00007fff328234be()	Unknown
 	ucrtbase.dll!00007fff32822b6f()	Unknown
 	vcruntime140.dll!00007fff222d202c()	Unknown
 	vcruntime140.dll!00007fff222d1b66()	Unknown
 	vcruntime140.dll!00007fff222dba10()	Unknown
 	ntdll.dll!00007fff353fa07d()	Unknown
 	ntdll.dll!00007fff35369c58()	Unknown
 	ntdll.dll!00007fff35365851()	Unknown
 	KernelBase.dll!00007fff31bc9e08()	Unknown
 	vcruntime140.dll!00007fff222d44f2()	Unknown
 	interface.exe!00007ff693b079d9()	Unknown
 	[External Code]	
 	Qt5Core.dll!0000000050dd882a()	Unknown
 	Qt5Core.dll!0000000050dd6aa1()	Unknown
 	Qt5Core.dll!0000000050de1d9b()	Unknown
 	interface.exe!00007ff693acf0f5()	Unknown
 	Qt5Core.dll!0000000050f9d8a5()	Unknown
 	Qt5Widgets.dll!0000000075bd97f0()	Unknown
 	Qt5Widgets.dll!0000000075bd87de()	Unknown
 	Qt5Core.dll!0000000050f78759()	Unknown
 	Qt5Core.dll!0000000050f7a5a1()	Unknown
 	qwindows.dll!00007ffeecb2394f()	Unknown
 	Qt5Core.dll!0000000050fc1429()	Unknown
 	[External Code]	
 	Qt5Core.dll!0000000050fc0aed()	Unknown
 	qwindows.dll!00007ffeecb23929()	Unknown
 	Qt5Core.dll!0000000050f747f2()	Unknown
 	Qt5Core.dll!0000000050f774a4()	Unknown
 	interface.exe!00007ff693b6b64e()	Unknown
 	interface.exe!00007ff6941cc234()	Unknown
 	interface.exe!00007ff6941929cb()	Unknown
 	[External Code]

**Comment by Menithal** [comment]
Bringing in partity to master just to eliminate variables for the crashing.

**Comment by Nex-Pro** [comment]
On hold awaiting new build.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3091/HighFidelity-Beta-PR11301-e1acf9dbc9a24c7a182296360f29ec0c1368f4ab.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3091/HighFidelity-Beta-PR11301-e1acf9dbc9a24c7a182296360f29ec0c1368f4ab.exe

**Comment by Nex-Pro** [comment]
recommencing.

**Comment by Nex-Pro** [comment]
On re-running this ( with the new build )  I created a cube to stand on. It was collisionless until I adjusted it in the Y axis! Didn't try this in the previous build.
This repro'd. Does not correct collisions when entity list is refreshed.
Will check against dev latest.


Test 1 **SUCCESS**

Test 2 **SUCCESS**

Crashing has gone away. 

Basic smoke test **SUCCESS**

**Comment by Menithal** [comment]
Saw the odd bug from Piper. 

Seemed to be very difficult to repo after client restart and most likely not related to this PR. Could not get a repo of it my self.

I'd say Pass with the PR Specific results Piper gave me.

**Comment by Nex-Pro** [comment]
after further examination this is a **PASS**

File: libraries/entities-renderer/src/RenderableModelEntityItem.cpp [hunk 1]
```diff
@@ -952,13 +956,21 @@ void ModelEntityRenderer::animate(const TypedEntityPointer& entity) {
...
-            }
+            } else if (!allowTranslation && index < animationJointNames.size()){
+                QString jointName = fbxJoints[index].name; // Pushing this here so its not done on every entity, with the exceptions of those allowing for translation
+                
+                if (originalFbxIndices.contains(jointName)) {
+                    // Making sure the joint names exist in the original model the animation is trying to apply onto. If they do, then remap and get it's translation.
+                    int remappedIndex = originalFbxIndices[jointName] - 1; // JointIndeces seem to always start from 1 and the found index is always 1 higher than actual.
+                    translationMat = glm::translate(originalFbxJoints[remappedIndex].translation);
+                }
+            }
```
Comment (AndrewMeadows) [inline libraries/entities-renderer/src/RenderableModelEntityItem.cpp]: This code performs fewer checks to achieve the same thing:
```
if (allowTranslation) {
    if (index < translations.size()) {
        translationMat = glm::translate(translations[index]);
    }
} else if (index < animationJointNames.size()) {
    QString jointName = fbxJoints[index].name;
    if (originalFbxIndices.contains(jointName)) {
        int remappedIndex = originalFbxIndices[jointName] - 1;
        translationMat = glm::translate(originalFbxJoints[remappedIndex].translation);
    }
}
```

File: libraries/entities-renderer/src/RenderableModelEntityItem.cpp [hunk 2]
```diff
@@ -909,8 +909,12 @@ void ModelEntityRenderer::animate(const TypedEntityPointer& entity) {
     QVector<JointData> jointsData;
 
     const QVector<FBXAnimationFrame>&  frames = _animation->getFramesReference(); // NOTE: getFrames() is too heavy
+    QStringList animationJointNames = _animation->getGeometry().getJointNames();
     auto& fbxJoints = _animation->getGeometry().joints;
+
     auto& originalFbxJoints = _model->getFBXGeometry().joints;
+    auto& originalFbxIndices = _model->getFBXGeometry().jointIndices;
```
Comment (AndrewMeadows) [inline libraries/entities-renderer/src/RenderableModelEntityItem.cpp]: Move the declaration and calculation of these new variables down closer to where they are actually used.  In particular, they should be AFTER the early exits below.

Comment (AndrewMeadows) [inline libraries/entities-renderer/src/RenderableModelEntityItem.cpp]: Move the declaration and calculation of these new variables down closer to where they are actually used. In particular, they should be AFTER the early exits below.

**Comment by Menithal** [comment]
Will get to it as soon as I get home from commute :)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3197/HighFidelity-Beta-PR11301-40d339a8318e7b014c60da12a6817aa684d44a9b.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3197/HighFidelity-Beta-PR11301-40d339a8318e7b014c60da12a6817aa684d44a9b.dmg

**Comment by Menithal** [comment]
@AndrewMeadows  Changes are done.

**Pull request merged**